### DISCUSSION TITLE
[Profiler] Explanation of the `Symfony initialization` time

### DISCUSSION THREAD

**Discussion by ToshY** [symfony/symfony#54341] [Q&A]
I cannot seem to find documentation regarding `Symfony initialization` in context of the profiler.

![image](https://github.com/symfony/symfony/assets/31921460/4ec66a4b-c1c4-4e78-b887-e774178b16e4)

The example screenshot above denotes `225 ms` initialization time (I have +/- `200 ms` on every page for this metric), but there's no graph or explanation on where the initialization time comes from.

So my questions are:

1. What constitutes as "Symfony initialization"?
2. What can I do to improve the performance / make it faster?
3. Is an initialization time of +/- `200 ms` regarded as "normal" for a standard Symfony application?

**Version**

- Symfony: `7.0.5`
- PHP: `8.3.3`

---

I've already checked the [Performance Checklists](https://symfony.com/doc/current/performance.html) and tried the followings settings, which doesnt improve the initialization time performance on dev (debug).
- Adding `.container.dumper.inline_factories: true`
- Adding `debug.container.dump: false`
- Adding OPcache settings

**Comment by MatTheCat** [comment]
https://github.com/symfony/symfony/blob/a9133af83ef26f0edaceca38aa885bb4f4d25a19/src/Symfony/Component/HttpKernel/DataCollector/TimeDataCollector.php#L102

To be more precise, it starts when the kernel is booted and ends when the `kernel.request` event is dispatched.

Most of the initialization time is spent initializing bundles and the container, so it does not really make sense to try improving it on your end: running your application on a fresh build (which should always happen on prod) should be enough to greatly reduce this time.

 I guess the hardware can make a significant impact too: on an old PC I have initializations taking more than 400ms without any cache, and 4ms when it is fresh.

**Reply by ToshY** [reply]
Hey @MatTheCat 👋 

Thanks for your answer. So basically I should not focus on trying to improve performance based on the initialization time, something which is basically out of my control.

**Reply by MatTheCat** [reply]
Yep unless it bogs down development, I wouldn’t worry about it.

**Reply by ToshY** [reply]
While I don't seem to have a performance impact in production, in hindsight my question still hasn't been answered, and I fail to find the rootcause.

> Most of the initialization time is spent initializing bundles and the container

If that's the case, why can't the debug toolbar specify which part causes slow initialization? I've been using [php-spx](https://github.com/NoiseByNorthwest/php-spx/) trying to figure it out, and besides the `Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher::dispatch` (which seems debug specific), I don't really see much slowing the application down.

A clean Symfony `6.4` demo application shows me an initializiation time of `44 ms`, so while I can agree it has something to do with either bundles/kernel/container, it's not easy to pinpoint.

**Reply by MatTheCat** [reply]
> why can't the debug toolbar specify which part causes slow initialization?

I guess this is because the profiler needs the container, so it cannot run before it is built?

Anyways if you cannot find what is slow using php-spx I doubt the profiler could help :confused:

**Reply by ToshY** [reply]
Hey @MatTheCat 

I'm coming back on this after a year because I still couldn't figure it out with the profiler, SPX or debugging in Symfony kernel itself. 

But that being said, I just couldn't believe it was really that "slow", so I thought of enabling opcache for development once again. I've found out 2 things:

1. I had a `opcache.ini` but apparently `opcache.enable` was never set to `1` (due to some docker environment variable misconfiguration 🤦‍♂️ ), so not only didn't opcache work on development, it was also never enabled on my production server 🙈.
2. Enabling opcache for development changed my initialization time from `+/- 225 ms` to `+/- 40-60 ms`. Safe to say production is also noticably faster now 🙃 

Thanks for the help.

**Reply by MatTheCat** [reply]
Ouch..! https://wiki.php.net/rfc/make_opcache_required should prevent this kind of issue in the future 🤞

**Comment by ntoniazzi** [comment]
Hi,
we have the same worries, except that in our case, the initialization time climbs up to **2200ms** in `dev` environment. It seems much smaller in `prod`, though I have no real metrics there. I'm running Symfony `5.4` on PHP `8.2`. I can see it in production, but unfortunately not locally. I may have an unwanted io, database check or funny bundle behavior, and it would help to have some intel on what's going on there.

Would you have any advice on how to investigate?

**Reply by MatTheCat** [reply]
> the initialization time climbs up to 2200ms in `dev` environment
> […]
> I can see it in production, but unfortunately not locally

You’re not really supposed to use the `dev` environment on production. If this is the only combination triggering a high initialization time, I would simply suggest to stop doing that :grin: 

Anyways, I guess if you wanted to profile you could do it with [Xdebug](https://www.jetbrains.com/help/phpstorm/profiling-with-xdebug.html) or using an APM like [Blackfire](https://www.blackfire.io/#profiling).

**Reply by alexandre-le-borgne** [reply]
Verify that PHP Jit is active
Try to run `composer --optimize --classmap-authoritative` or `composer -o -a`
 (Cf https://getcomposer.org/doc/articles/autoloader-optimization.md)
 Try to Profile with Xdebug 
 Exemple of integration  with PHPStorm:  https://www.jetbrains.com/help/phpstorm/profiling-with-xdebug.html#enable-profiling-with-xdebug

**Reply by ntoniazzi** [reply]
>  I would simply suggest to stop doing that

Well, that's a sensible suggestion, I think I'll stick to that :)
I'll give xdebug and php-spx a try, though.

Thanks, @alexandre-le-borgne and @MatTheCat .

**Reply by ntoniazzi** [reply]
I played a little with SPX and found out that a lot of time was spent on **loading files** during "Symfony initialization".
It turns out that on our servers, the [open_basedir](https://www.php.net/manual/en/ini.core.php#ini.open-basedir) ini directive had been activated, **disabling the realpath cache**! Every file lookup thus took more time.
The difference between prod and dev environments being that in dev, every service is stored in its own php file, rather that gathered in a single container file.
We disabled the `open_basedir` and everything went back to normal.
